### PR TITLE
[SPARK-38549][SS] Add `numRowsDroppedByWatermark` to `SessionWindowStateStoreRestoreExec`

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch adds `numRowsDroppedByWatermark` metric to `SessionWindowStateStoreRestoreExec` operator.

### Why are the changes needed?

`SessionWindowStateStoreRestoreExec` filters out outdated inputs by watermark but it doesn't provide corresponding `numRowsDroppedByWatermark` metric. It makes users confused about stream query result.

### Does this PR introduce _any_ user-facing change?

Yes, adding a metric.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Thanks for the fix! 

Would you mind adding the actual tests or steps you've gone through manual tests?
And given state store write is the operator in charge to report the metrics in streaming listener, would it be coupled with the metric "numRowsDroppedByWatermark" in SessionWindowStateStoreSaveExec? If not, shall we spend time a bit to see whether it is feasible or not?

Thanks in advance!

**Comment by viirya** [comment]
We know `SessionWindowStateStoreSaveExec` is behind `SessionWindowStateStoreRestoreExec` in the operator order. So if input rows are dropped by `SessionWindowStateStoreRestoreExec`, we won't see them in later operators such as `SessionWindowStateStoreSaveExec`.

That's why we observed that some rows seems dropped by watermark, but we don't see any `numRowsDroppedByWatermark`.

`SessionWindowStateStoreRestoreExec` is not a state store writer, so it doesn't have `numRowsDroppedByWatermark` metric, but it actually drops input rows by watermark predicate. It is confused to end users as they cannot accurately measure the number of dropped by watermark.

**Comment by viirya** [comment]
cc @dongjoon-hyun

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @viirya .

**Comment by HeartSaVioR** [comment]
My point is that the operator `SessionWindowStateStoreSaveExec` would not still produce the late events being dropped in `SessionWindowStateStoreRestoreExec ` for streaming listener, hence this is going to be a "partial" fix. I'm asking to see whether we have a chance to fix the issue holistically while we are here and you see the problem.

**Comment by HeartSaVioR** [comment]
For example, we have a test covering the functionality of checking the number of dropped events in EventTimeWatermarkSuite via https://github.com/apache/spark/blob/1d4e917d5c36ed53d5d97d28e2a16cb2f9a3b81c/sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala#L854-L866

This does not work with session window aggregation, and this PR does not fix the issue if I understand correctly. I won't insist if you mind spending more time seeking for the possibility.

**Comment by viirya** [comment]
Sorry, let me rephrase your words...I need to re-read it a few times to know what you mean.

`StateOperatorProgress` is only created for `StateStoreWriter` in `ProgressReporter`. So if we need to have `StateOperatorProgress` for `SessionWindowStateStoreRestoreExec`, otherwise it will only show in Spark UI.

It makes sense.

Is this what you mean?

**Comment by viirya** [comment]
If so, how about I merge this metric/UI fix and work on the another issue?

**Comment by HeartSaVioR** [comment]
> StateOperatorProgress is only created for StateStoreWriter in ProgressReporter. So if we need to have StateOperatorProgress for SessionWindowStateStoreRestoreExec, otherwise it will only show in Spark UI.

Yes, and accumulated number of late events in SS UI page is also based on StateOperatorProgress. This PR only adds the new metric in SessionWindowStateStoreRestoreExec (as SQL metric) which is not connected with the SS metric mechanism.

> If so, how about I merge this metric/UI fix and work on the another issue?

Yeah please feel free to do so. Thanks for taking the step!

**Review by HeartSaVioR — APPROVED** [review]
+1 assuming we'll give a try to fix the other issue on metric soon.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun @HeartSaVioR

**Comment by viirya** [comment]
Merging to master.
